src/main/java/net/minecraftforge/client/model/CompositeModel.java

```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    public static class Geometry implements IMultipartModelGeometry<Geometry>
+    {
+        private final ImmutableMap<String, Submodel> parts;
+
+        Geometry(ImmutableMap<String, Submodel> parts)
+        {
+            this.parts = parts;
+        }
+
+        @Override
+        public Collection<? extends IModelGeometryPart> getParts()
+        {
+            return parts.values();
+        }
+
+        @Override
+        public Optional<? extends IModelGeometryPart> getPart(String name)
+        {
+            return Optional.ofNullable(parts.get(name));
+        }
+
+        @Override
+        public IBakedModel bake(IModelConfiguration owner, ModelBakery bakery, Function<ResourceLocation, TextureAtlasSprite> spriteGetter, ISprite sprite, VertexFormat format, ItemOverrideList overrides)
+        {
+            String particleLocation = owner.resolveTexture("particle");
+            TextureAtlasSprite particle = spriteGetter.apply(new ResourceLocation(particleLocation));
+
+            ImmutableMap.Builder<String, IBakedModel> bakedParts = ImmutableMap.builder();
+            for(Map.Entry<String, Submodel> part : parts.entrySet())
+            {
+                Submodel submodel = part.getValue();
+                if (!owner.getPartVisibility(submodel))
+                    continue;
+                bakedParts.put(part.getKey(), submodel.bake(bakery, spriteGetter, sprite, format));
+            }
+            return new CompositeModel(owner.isShadedInGui(), owner.useSmoothLighting(), particle, bakedParts.build(), owner.getCombinedState(), overrides);
+        }
+
+        @Override
+        public Collection<ResourceLocation> getTextureDependencies(IModelConfiguration owner, Function<ResourceLocation, IUnbakedModel> modelGetter, Set<String> missingTextureErrors)
+        {
+            Set<ResourceLocation> textures = new HashSet<>();
+            for(Submodel part : parts.values())
+            {
+                textures.addAll(part.getTextureDependencies(owner, modelGetter, missingTextureErrors));
+            }
+            return textures;
+        }
+    }
+
+    public static class Loader implements IModelLoader<Geometry>
+    {
+        public static final Loader INSTANCE = new Loader();
+
+        private Loader() {}
+
+        @Override
+        public void onResourceManagerReload(IResourceManager resourceManager)
+        {
+
+        }
+
+        @Override
+        public Geometry read(JsonDeserializationContext deserializationContext, JsonObject modelContents)
+        {
+            if (!modelContents.has("parts"))
+                throw new RuntimeException("Composite model requires a \"parts\" element.");
+            ImmutableMap.Builder<String, Submodel> parts = ImmutableMap.builder();
+            for(Map.Entry<String, JsonElement> part : modelContents.get("parts").getAsJsonObject().entrySet())
+            {
+                // TODO: Allow customizing state? If so, how?
```

Comment:
Ah I should get rid of this comment. The "how" answered itself a while ago through the "transform" feature.
You can do something like this:
```json
{
"loader": "forge:composite",
"parts": {
  "part1: {
    "parent": x,
    "transform": { "rotation": ..., "translation": ..., "scale" }
  }
}
```

Reply:
It might still be worthwhile to support the much simpler vanilla-style x/y rotation, since that can be trivially converted to a TRSR by existing code.

Reply:
That would mean adding even more values to the model deserializer, or creating a dedicated deserializer for this. It's currently parsed as if it was any other model json file.